New: Add context.report({ messageId }) (fixes #6740)

j-f1 opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

Add something to the core

Fixes #6740.

<!--
    Please ensure your pull request is ready:
-->

- [x] Read the pull request guide (http://eslint.org/docs/developer-guide/contributing/pull-requests)
- [x] Include tests for this change
- [x] Update documentation for this change (if appropriate)

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**
I added a new `messageId` parameter to `context.report`:

```js
context.report({
  messageId: 'foo'
})
```

This takes the value `foo` key from `meta.messages`:

```js
module.exports = {
  meta: {
    // ...
    messages: {
      foo: "Your message here"
    }
  },
  create(context) {
    // ...
  }
}
```

You can also use them in `RuleTester` tests:

```js
ruleTester.run('rule-name', rule, {
  valid: [
    // ...
  ],
  invalid: [
    {
      code: 'Something here',
      errors: [
        {
          messageId: 'foo'
          // if you have {{ }} markers in your message, add a `data` key here:
          // data: {
          //   myKey: 'my value'
          // }
        }
      ]
    }
  ]
})
```

---

If my example (`accessor-pairs`) looks good, I’ll be happy to convert some more rules.

platinumazure commented:
Nice! This looks pretty good at first glance, I'll give it a proper review later. Thanks for taking the time to write this- I'm looking forward to this getting in.

not-an-aardvark commented on lib/rule-context.js:
```diff
@@ -185,6 +185,19 @@ class RuleContext {
+                if (!messages || Object.keys(messages).indexOf(id) === -1) {
+                    throw new TypeError(`context.report() called with a messageId of '${id}' which is not present in the 'messages' config: ${JSON.stringify(messages, null, 2)}`);
+                }
+                descriptor.message = messages[id];
```
I see we're only using `messageId` in `RuleContext`, and it gets replaced with `message` by the time it reaches formatters. Do we want to preserve `messageId` in formatters? That would allow formatters to do i18n by swapping out messages based on the `messageId`, which I thought was a goal of the proposal.

(That would also make the `messageId` part of a rule's public API since changing it could break external formatters. This may or may not be desirable.)

(Also, it might be difficult for formatters to replace messages without having access to the `data` from the rule, since it won't be possible to unambiguously parse interpolated messages to determine what the interpolated data is.)

j-f1 replied:
I was thinking that rules would do something like

```js
messages: {
  foo: _('Don’t do foo!')
}
```

not-an-aardvark commented on lib/rule-context.js:
```diff
@@ -185,6 +185,19 @@ class RuleContext {
 
+            if (descriptor.messageId) {
+                const id = descriptor.messageId;
+                const messages = this.meta.messages;
```
I think this will throw a TypeError if a rule doesn't have a `meta` property.

platinumazure commented:
@not-an-aardvark Regarding internationalization, don't you think it might make more sense for rules to own the localization process? At any rate, I'm not sure formatters should be responsible for it... At least, that was what I had thought originally.

not-an-aardvark commented:
@platinumazure I was thinking maybe formatters would be a better solution because they would solve the issue of "how to configure i18n" (it would be configured by simply selecting a formatter for the user's language). While the current API is a step in the right direction, as far as I can tell it doesn't appear to allow any configuration of message mappings, which would be necessary for i18n.  Also, delegating i18n to formatters might help take the burden off of the authors of popular plugins, so they wouldn't have to maintain many different translations in addition to the rule code itself.

Having said that, I don't have a perfect solution to the issue, given that passing messages to formatters before doing interpolation might require a significant API change. So I'm not opposed to this implementation -- I just want to make sure we've considered the details if we're planning to use this as a way to do i18n.

j-f1 commented:
This is now mergeable again @not-an-aardvark. Can you take another look at it?

not-an-aardvark commented on lib/report-translator.js:
```diff
@@ -261,11 +263,29 @@ module.exports = function createReportTranslator(metadata) {
+            if (!messages || !Object.prototype.hasOwnProperty.call(messages, id)) {
+                throw new TypeError(`context.report() called with a messageId of '${id}' which is not present in the 'messages' config: ${JSON.stringify(messages, null, 2)}`);
+            }
+            descriptor.message = messages[id];
```
Can you avoid mutating the descriptor here? I think it would be better to put the `messageId` logic in `interpolate` so that all of the message-handling logic is in the same place, and the message doesn't need to be reassigned.

(Unlike https://github.com/eslint/eslint/pull/9165#discussion_r135395956, this time I'm not concerned about mutating objects from rules, since I see you cloned the descriptor above. I just think not mutating a function's arguments in general is a good idea for code cleanliness, since it makes code easier to follow.)

j-f1 replied:
Hm, this isn’t actually a function argument:

```js
// lib/report-translator.js:262
const descriptor = normalizeMultiArgReportCall.apply(null, arguments);
````

not-an-aardvark replied:
Oops, you're right. I still think it would be slightly better to avoid mutation, but I don't feel strongly about it.

not-an-aardvark commented on lib/report-translator.js:
```diff
@@ -216,6 +211,12 @@ function createProblem(options) {
+    // If this isn’t in the conditional, some of the tests fail
+    // because `messageId` is present in the problem object
+    if (options.messageId) {
+        problem.messageId = options.messageId;
```
I'm confused -- do we want to include the `messageId` in the problem, or not? (Also cc'ing @platinumazure, since this is more of a design question)

Sorry if it seems like I'm going back and forth on this. This change adds some complexity to the rule API, and it will be difficult to make modifications to the API later without breaking compatibility. So I would feel better about the change knowing that (a) it provides benefits that clearly outweigh the additional complexity costs, and (b) we've thoroughly discussed the design details to ensure that those benefits will actually be accessible with the API that we expose.

I'm assuming you added `problem.messageId` in response to my suggestion about integrations in https://github.com/eslint/eslint/pull/9165#discussion_r135396568. Thanks for doing that, although I'm unsure whether it's useful for an integration to have access to the `messageId` if they don't also have access to the data that was used as part of message interpolation, since that data would be needed to reconstruct the message or use i18n.

I'm a bit unclear about what the design goal of this feature is (maybe @platinumazure can help explain). From what I can tell (feel free to correct me if I describe anything incorrectly), one of the following is the case:

1. We want i18n to be handled by plugins and rule authors. In that case, it seems like message IDs are an internal detail and don't need to be exposed as part of the final problem object. (I'm unsure whether `messageIds` provide a significant a benefit over manual string concatenation in this case, although there might be something I'm missing.)
1. We want i18n to be handled by integrations. In that case, we would probably need to let the integrations handle message interpolation as well, otherwise they won't have enough info to translate messages.

I'm not trying to stand in the way here -- I just want to make sure we're communicating the design goals clearly, so that the resulting API is coherent and useful.

j-f1 replied:
How does this sound?

1. Remove the `messageId` from the report
2. Open an issue to allow overriding the message-generating infrastructure
3. Allow the values in the `messages` object to be functions that get called to retrieve the value. Usage:
    ```js
    module.exports = {
      meta: {
        messages: {
          foo: (optionalDataParam) => myI18nLib('message here')
        }
      }
    }
    ```

not-an-aardvark replied:
That seems like it could work, although I'd also like to hear what @platinumazure thinks.

j-f1 replied:
@platinumazure Any comments on this design?

kaicataldo replied:
@platinumazure Friendly ping

platinumazure replied:
Ack, I didn't see this thread when I commented earlier today. Something like the proposed design could also work, but I'm a fan of some sort of core-level hook that would allow i18n/l10n to be separate from rules, if desired.

j-f1 commented:
I’d also appreciate feedback on the IDs I’ve used in the rules I upgraded.

j-f1 commented:
I’m not sure why the AppVeyor build failed. Closing and reopening to restart the build.

j-f1 commented:
Still failing. Any ideas why @not-an-aardvark?

not-an-aardvark commented:
It seems to be unrelated to this PR -- the last two builds on master have failed too ([commits](https://github.com/eslint/eslint/commits/master)).

not-an-aardvark commented:
@j-f1 I think the issue is fixed now (see https://github.com/eslint/eslint/issues/9307)

j-f1 commented:
@not-an-aardvark Any thoughts on this?

not-an-aardvark commented:
My thoughts are pretty much the same as they were in https://github.com/eslint/eslint/pull/9165#discussion_r136728029.

Sorry about the delay here -- I realize it's frustrating for a PR to not get merged for awhile, and I'm hoping we can resolve this soon. That said, whatever we decide to do here will effectively need to be supported indefinitely, so I think it's worth taking the time to get the details right.

platinumazure commented:
I'm really sorry, work has nearly killed me but I'm not dead yet. Hoping to address questions this week... I greatly appreciate everyone's patience and I want to thank @j-f1 for sticking with this. Really appreciate it.

platinumazure commented:
Work continued to twist the knife over the last few weeks... Really sorry.

I was thinking about the i18n dilemma. Here's my rough thinking on it:

1. Actually designing a solid i18n mechanism that would work for everyone is probably beyond the scope of this change :smile:
1. Rules/plugins themselves are a pretty good place to maintain i18n and l10n, if the author desires. For plugin authors, this solution would allow them to immediately start internationalizing and localizing their messages if desired.
1. Integrations/formatters are probably not a good place to maintain l10n, because a formatter could not be expected to know about plugins unrelated to it.
1. An ideal solution would allow for the possibility of injecting l10n from somewhere else, as a separate entity that is neither an integration/formatter nor the rule/plugin itself.

Here's an example of what I have in mind for the last (but again, probably beyond the scope of this PR):

* ESLint core would create a message internationalization hook which would allow users to specify a `messageLocalizers` (name open for bikeshedding) array in configuration. Each of the `messageLocalizers` would receive the rule name and messageId, and would be responsible for looking up an internationalized version of the message or just returning the original message.
    * Best case would be if the message localizer could just return the message template and ESLint could hydrate it with the data from the original problem report, to ensure separation of concerns.
    * For plugins that are happy to supply their own i18n, maybe they could also export a messageLocalizer that could be set in config?
* Each of the `messageLocalizers` would be tried in order, and the first one that can localize the message would include a property indicating this success in its response.

I think with a design like that, it would be possible to allow for one or more full internationalization/localization libraries to develop that supports localizing rules in whatever languages they choose to support, for whatever set of rules/plugins they choose to support.

Personally, I think it would be okay to avoid trying to solve the i18n problem in this PR.

@not-an-aardvark @j-f1 Any thoughts on this?

j-f1 commented:
+💯 on not adding direct support for i18n in this PR — that can already be covered in the rules like you mentioned.

Should I try to update more of the core rules in this PR, or should I do that in another PR?

platinumazure commented:
@j-f1 My suggestion would be to definitely do those other rules separately. In fact, even doing one rule feels a little much to me, but then again it's good to prove the concept. :smile:

I'll check @not-an-aardvark's review and see if it's worth me doing a review on top or not.

platinumazure commented on docs/developer-guide/working-with-rules.md:
````diff
@@ -181,6 +181,60 @@ Note that leading and trailing whitespace is optional in message parameters.
+  ],
+})
+```
+
````
Do we need an example of representing an error message with data in RuleTester?

j-f1 replied:
✅

platinumazure commented on lib/report-translator.js:
```diff
@@ -41,7 +42,10 @@ function normalizeMultiArgReportCall() {
-        return arguments[0];
+
+        // Shallow clone the object to avoid surprises
+        // if reusing the descriptor
```
We've recently enabled a rule in our codebase related to comments spanning multiple lines. I would suggest rebasing and rerunning `npm run lint` to be sure this won't be flagged.

not-an-aardvark commented:
Thanks @platinumazure, that makes sense to me. I have a few smaller questions/notes:

* Does using `messageId` change a rule's backwards-compatibility guarantees? In the past, we've generally been able to change a rule's message in a semver-patch release. But if a rule was being used with an external `messageLocalizer`, it seems like the messages and placeholders would become part of the rule's public API. For example, if a rule author decides to remove a piece of information from a message, they would need to continue supplying the corresponding `data` property to avoid breaking other localizers.

    As an alternative, we could use the default set of localizers as a fallback if either (a) there is no available substitute message for the given id, or (b) the provided substitute message is missing a required `data` placeholder.

* I think modifying core rules to replace `message` with `messageId` would be a breaking change for [`eslint-rule-composer`](https://github.com/not-an-aardvark/eslint-rule-composer), which runs ESLint rules on its own and assumes that `context.report` calls have have a `message` property. I would be fine with updating `eslint-rule-composer`, but it's possible that there are other integrations that would break as well. I bring this up because to some extent, the behavior of a rule is part of our public API because we expose core rules with `Linter#getRules`.

    We could have a policy that the behavior of a core rule is only well-defined when the rule is run by the ESLint version that it ships with (as opposed to other libraries like `eslint-rule-composer` that could run the rule), although this would effectively make `eslint-rule-composer` impossible to use without relying on a private API.

* With regard to this PR: I think it would actually be slightly better if we only implement the core changes in this PR, and then we update rules afterwards. As-is, it's a bit difficult to review the core changes here because there are so many other file changes.

j-f1 commented:
> With regard to this PR: I think it would actually be slightly better if we only implement the core changes in this PR, and then we update rules afterwards. As-is, it's a bit difficult to review the core changes here because there are so many other file changes.

👍
Should I keep the `no-alert` change in this PR or extract it, too?

not-an-aardvark commented:
I suppose it would be better to extract that too, but I don't feel strongly about it either way.

j-f1 commented:
See [`use-messageids`](https://github.com/j-f1/forked-eslint/tree/use-messageids) in my fork for the rule refactoring commits.

platinumazure commented:
@not-an-aardvark You raise a lot of good questions... I don't really have any good answers yet. Sorry, need a little more time to think it through.

platinumazure commented:
@not-an-aardvark Okay, I have some rough thoughts, let me know what you think:

> * Does using `messageId` change a rule's backwards-compatibility guarantees? In the past, we've generally been able to change a rule's message in a semver-patch release. But if a rule was being used with an external `messageLocalizer`, it seems like the messages and placeholders would become part of the rule's public API. For example, if a rule author decides to remove a piece of information from a message, they would need to continue supplying the corresponding `data` property to avoid breaking other localizers.

Yes, I think I agree with this. And I'm no longer 100% sure this is worth doing as a result (might need to discuss with team or TSC in chat to flesh that out).

My proposal for how to integrate rule messages into the semver policy:

* Converting a rule from `message` to `messageId` is semver-minor since it's an update of how we present the rules.
* Changing a message without changing the number or name of message placeholders is semver-patch.
* Changing a message and adding a message placeholder is semver-minor.
* Changing a message to remove or rename a message placeholder is semver-major. If the old placeholder continues to be emitted by the rule, it could be done in a semver-minor release, with the removal of the old placeholder being semver-major.

>   As an alternative, we could use the default set of localizers as a fallback if either (a) there is no available substitute message for the given id, or (b) the provided substitute message is missing a required `data` placeholder.

I agree that we could have the "default localizer" be a fallback in case of unsuccessful localizing due to renamed/removed placeholder (where "default localizer" basically means the current message placeholder logic). I think ideally, I would prefer to see this implemented as a CLIEngine/CLI option, though, so that users can opt into it if they want rather than having ESLint only use localizers set up in configuration.

> * I think modifying core rules to replace `message` with `messageId` would be a breaking change for [`eslint-rule-composer`](https://github.com/not-an-aardvark/eslint-rule-composer), which runs ESLint rules on its own and assumes that `context.report` calls have have a `message` property. I would be fine with updating `eslint-rule-composer`, but it's possible that there are other integrations that would break as well. I bring this up because to some extent, the behavior of a rule is part of our public API because we expose core rules with `Linter#getRules`.
> 
>   We could have a policy that the behavior of a core rule is only well-defined when the rule is run by the ESLint version that it ships with (as opposed to other libraries like `eslint-rule-composer` that could run the rule), although this would effectively make `eslint-rule-composer` impossible to use without relying on a private API.

I'm jumping ahead of the scope of this PR, but could we expose a module `MessageLocalizer` with method `normalizeMessage` which could take in a message or messageId and data, and run the same logic that ESLint would use to try to localize/format messages (i.e., go through all user-specified localizers, possibly fall back to default logic), in such a way that eslint-rule-composer wouldn't need to know if it got a hydrated message, an un-hydrated message + data, or a messageId + data?

Let me know what you think.

j-f1 commented:
@not-an-aardvark and @platinumazure Can you take another look when you have a chance? I think I’ve addressed all of the issues you’ve brought up (except for i18n support in formatters, but that’s a whole different can of worms).

ilyavolodin commented:
@not-an-aardvark @platinumazure Could you review this again please? It's been sitting open for a very long time, and looks pretty good to me.

j-f1 commented:
@platinumazure Friendly ping?

platinumazure commented:
Hi @j-f1, I do apologize. I still need to chat with @not-an-aardvark about the semver/API compatibility implications he raised. This PR will make message changes backward-incompatible and we need to figure out if that's something we can safely allow, and/or if there's anything we can do to avoid this.

I'm sorry I haven't had a chance to figure this out sooner.

j-f1 commented:
Proposal from [above](https://github.com/eslint/eslint/pull/9165#discussion_r136797596):

> How does this sound?
> 
> 1. Remove the `messageId` from the report
> 2. Open an issue to allow overriding the message-generating infrastructure
> 3. Allow the values in the `messages` object to be functions that get called to retrieve the value. Usage:
>     ```js
>     module.exports = {
>       meta: {
>         messages: {
>           foo: (optionalDataParam) => myI18nLib('message here')
>         }
>       }
>     }
>     ```
>

platinumazure commented:
@j-f1 How would a message overriding infrastructure look at a high level? If there is no `messageId`, then would message overriding tools be passed the un-hydrated message or the hydrated message? How would this address the semver concerns brought up by @not-an-aardvark?

j-f1 commented:
> Do any further changes need to be made to this PR to pivot in the suggested direction, or do you think this is ready to merge?

None that I can think of, so this should be good to go.
